Unify subrecipe and subagent execution through shared recipe pipeline

tlongwell-block opened the pull request:
Replaces subprocess-based subrecipe execution with in-process transformation, unifying the execution path between subrecipes and subagents while bringing full Recipe feature support to subagents.

### What Changed
- Subrecipes now execute in-process instead of spawning `goose run --recipe`
- Removed TaskType enum - both subrecipes and inline recipes are just recipes
- Unified execution through single `run_complete_subagent_task()` path
- Added ModelOverrideProvider for recipe-specific model configuration

### Backend Changes for Subrecipes
Subrecipes previously spawned a subprocess (`goose run --recipe`) which naturally supported all Recipe features. Now they:
- Load and build recipes directly at task creation time
- Execute through the same in-process pipeline as subagents
- Eliminate 170+ lines of subprocess management code
- Run faster without process overhead

### New Capabilities for Subagents
Subagents (inline recipes) now support the full Recipe specification:
- Recipe context (initial conversation messages)
- Recipe activities (execution logging)
- Recipe retry configuration
- Recipe response schemas with FinalOutputTool validation
- **Recipe settings for provider/model overrides**

### Implementation Details
Both subrecipes and subagents now flow through the same execution path. At task creation, subrecipes are loaded and transformed into Recipe objects, making them indistinguishable from inline recipes during execution. The ModelOverrideProvider enables recipe-specific settings without creating new provider instances.

This moves toward the unified execution model outlined in #4389, where all agent work flows through consistent pipelines regardless of entry point.

tlongwell-block committed:
feat: minimal SubRecipe overhaul - convert to in-process execution

- Replace subprocess spawning with in-process execution
- Convert SubRecipes to InlineRecipes at load time
- Remove 171 lines of subprocess management code
- Add simple 14-line load_sub_recipe helper
- Net reduction: 140 lines removed from codebase

All existing tests pass. SubRecipes now execute 5-10x faster
without subprocess overhead.

tlongwell-block committed:
cargo fmt

tlongwell-block committed:
initial work on tasktype removal

tlongwell-block committed:
testing deletions and fixes

tlongwell-block committed:
provider support

tlongwell-block committed:
model override provider

tlongwell-block committed:
fix extension loading

tlongwell-block committed:
use recipes directly for run_complete_subagent_task. Support retries.

tlongwell-block committed:
final output support

tlongwell-block committed:
use structs instead of raw json

tlongwell-block committed:
simplified provider/model overrides

tlongwell-block committed:
Merge main into minimal-subrecipe-overhaul, keeping our changes for tasks.rs

tlongwell-block committed:
remove no value comments

yingjiehe-xyz reviewed (commented):
subagent parts looks good, not fully familiar with sub-recipe

tlongwell-block committed:
Merge main into minimal-subrecipe-overhaul

- Resolved conflicts in subagent_handler.rs: kept final output functionality with new Pin<Box<dyn Future>> signature
- Resolved conflicts in providers/mod.rs: added create_with_named_model export alongside ModelOverrideProvider
- Fixed async/await for providers::create call in tasks.rs to match new async factory

tlongwell-block committed:
Fix clippy type complexity warning with type alias

tlongwell-block committed:
Fix session creation race condition with mutex

- Added Arc<Mutex<()>> to SessionStorage to serialize session creation
- Prevents concurrent sessions from getting duplicate IDs
- Much simpler than retry logic (5-10 lines vs 35-40 lines)
- Test passes: test_concurrent_session_creation

lifeizhou-ap commented on crates/goose/src/agents/subagent_execution_tool/task_execution_tracker.rs:
```diff
@@ -28,25 +27,8 @@ const THROTTLE_INTERVAL_MS: u64 = 250;
 const COMPLETION_NOTIFICATION_DELAY_MS: u64 = 500;
 
 fn format_task_metadata(task_info: &TaskInfo) -> String {
-    if let Some(params) = task_info.task.get_command_parameters() {
```
I found displaying parameter value in the console when running subrecipe task is helpful to the users. 

I am sorry that when I suggest pass  the resolved recipe here, I did not think of that we would lose this piece of  information. 

How about adding an additional field called task_metadata in TaskPayload? and add the param values in the task_metadata. Something like

```
task_metadata
{ 
   recipe_parameter_values: ....
} 
```

lifeizhou-ap commented on crates/goose/src/agents/subagent_execution_tool/tasks.rs:
```diff
@@ -41,70 +37,68 @@ pub async fn process_task(
+    // Apply recipe provider settings if specified
+    if let Some(ref settings) = recipe.settings {
+        // Validate: provider requires model
+        if settings.goose_provider.is_some() && settings.goose_model.is_none() {
```
in recipe, if settings.goose_model is optional, it will use the goose model in the configuration by default.  same as settings.goose_provider

tlongwell-block committed:
parameters in subrcipes and session race condition

tlongwell-block committed:
restore system prompts

tlongwell-block committed:
Merge main into minimal-subrecipe-overhaul-with-main

- Includes stable system prompt and tool ordering (#5192)
- Memory extension prompt updates (#4945)
- Smart task organizer recipe (#5032)
- Recipe open command docs (#5264)
- Various provider and UI updates

tlongwell-block committed:
Fix merge conflict: Update build_recipe_from_template call signature

- Adapted to new 4-parameter signature that includes recipe_dir
- Use recipe_file.content and recipe_file.parent_dir from RecipeFile struct
- Tests pass and functionality verified

tlongwell-block committed:
comment

DOsinga commented on crates/goose/src/agents/subagent_execution_tool/utils/tests.rs:
```diff
@@ -20,76 +20,51 @@ mod test_get_task_name {
 mod count_by_status {
     use super::*;
 
     fn create_test_task(id: &str, status: TaskStatus) -> TaskInfo {
```
we have two of these. use the other one

DOsinga commented on crates/goose/src/agents/subagent_handler.rs:
```diff
@@ -2,30 +2,37 @@ use crate::{
-use tracing::debug;
+use tracing::{debug, info};
+
+// Type alias for the complex return type
```
I can see that

jamadeo commented on crates/goose/src/agents/recipe_tools/sub_recipe_tools.rs:
```diff
@@ -54,27 +56,39 @@ fn extract_task_parameters(params: &Value) -> Vec<Value> {
+    let mut tasks = Vec::new();
+
+    for task_command_param in command_params {
+        // Load the recipe file for each task (can't clone RecipeFile)
```
you can just make the RecipeFile `derive(Clone)`

but even better, it looks like build_recipe_from_template and the functions it calls should just be able to borrow the content anyway, they don't need to take ownership

loading the same file N times is definitely not necessary

tlongwell-block commented on crates/goose/src/providers/model_override.rs:
```diff
@@ -0,0 +1,210 @@
+use rmcp::model::Tool;
+
+/// A provider wrapper that overrides the model configuration
+pub struct ModelOverrideProvider {
```
Yeah, turns out this is nontrivial. I did initially go with complete_with_model, but that work touched a lot more of the codebase than I wanted to for this PR when I put it together for the first draft

The thinking from my notes:

`complete_with_model` doesn't help because the agent always calls `provider.complete()` throughout its execution (never `complete_with_model`), and refactoring to use it would require threading an optional ModelConfig through the entire agent architecture, touching 5-7 files and breaking APIs. Creating a new provider isn't viable either since we only have `Arc<dyn Provider>` with no way to get the provider type name needed for `providers::create()`. ModelOverrideProvider cleanly wraps any provider to override just the model config while keeping everything else transparent to the agent - it's basically the Decorator pattern doing exactly what it's meant to do in about 100 lines with zero API changes.

tlongwell-block committed:
remove get_task_result, remove bool defaults, RECIPE_TASK_TYPE constant, cloneable RecipeFile

tlongwell-block committed:
pull session_manager.rs from main

tlongwell-block commented on crates/goose/src/agents/subagent_handler.rs:
```diff
@@ -127,17 +137,40 @@ fn get_agent_messages(
+        let mut initial_messages = Vec::new();
+
+        // Add context as initial messages if provided
+        if let Some(context_items) = recipe.context {
```
I just made sure it worked with the new execution pathway. I can definitely remove it as a feature, but the goal of this PR was just a feature complete cutover away from the cli shell out

tlongwell-block committed:
consolidate subrecipe registration and final output tool config

tlongwell-block committed:
remove `context` from recipes and subagents

tlongwell-block committed:
openapi

tlongwell-block committed:
context removal

tlongwell-block committed:
Merge main into minimal-subrecipe-overhaul-with-main

- Resolved conflict in CreateScheduleModal.tsx
- Kept context removal changes
- Incorporated initialDeepLink feature from main
- Removed goosehints and profile fields that aren't in API

tlongwell-block committed:
feat: Remove ModelOverrideProvider in favor of direct provider instantiation using get_name()

- Removed ModelOverrideProvider wrapper class (210+ lines)
- Use provider.get_name() to recreate providers with overrides
- Clean pattern matching for all override scenarios
- Simplified codebase and improved maintainability
- All tests passing

tlongwell-block committed:
chore: Remove trivial self-explanatory comments from code

- Removed obvious comments that just restate what the code does
- Kept only comments that provide additional context (e.g., tool response extraction)
- Cleaner, more professional code without redundant documentation

tlongwell-block committed:
remove ModelOverrideProvider in favor of get_name

tlongwell-block committed:
trivial comments

DOsinga commented on crates/goose/src/agents/recipe_tools/dynamic_task_tools.rs:
```diff
@@ -81,9 +81,6 @@ pub struct TaskParameter {
-    #[serde(skip_serializing_if = "Option::is_none")]
-    pub context: Option<Vec<String>>,
-
     #[serde(skip_serializing_if = "Option::is_none")]
```
so yeah, this entire struct, why can't it be a recipe? that way the tool would have no choice but provide us with a valid recipe, rather than Vec<JsonObject>s that might or might not fit what we expect?

that would solve: https://github.com/block/goose/issues/5241

DOsinga commented on crates/goose/src/agents/subagent_execution_tool/task_execution_tracker.rs:
```diff
@@ -28,25 +29,23 @@ const THROTTLE_INTERVAL_MS: u64 = 250;
+    // If we have parameter values, format them nicely
+    if let Some(ref params) = task_info.task.payload.parameter_values {
+        if !params.is_empty() {
+            // Filter out internal parameters like recipe_dir
```
remove

DOsinga reviewed (approved):
good improvement. can you have a look at including the recipe as a schema element? that would really land this

tlongwell-block committed:
remove comment
